### PR TITLE
docs: [FE-269] Add documentation detailing configuration steps to set the values for ngpus.

### PR DESCRIPTION
**Pull request by jagadeesh545**
## Description

[FE-269](https://hpe-aiatscale.atlassian.net/browse/FE-269?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ)
DeterminedAI cannot function properly on an HPC cluster with PBS if the ngpus resource is not configured properly. Add instructions to the PBS Requirements section explaining how to configure the ngpus resource.

## Test Plan
Tested manually.
Screenshot 2024-01-18 at 22 34 34
Screenshot 2024-01-18 at 22 34 43
Screenshot 2024-01-18 at 22 34 53

## Checklist

- [x] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket
[FE-269](https://hpe-aiatscale.atlassian.net/browse/FE-269?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ)

[FE-269]: https://hpe-aiatscale.atlassian.net/browse/FE-269?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[FE-269]: https://hpe-aiatscale.atlassian.net/browse/FE-269?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: docs/setup-cluster/slurm/slurm-requirements.rst [hunk 1]
```diff
@@ -228,6 +228,94 @@ interacts with PBS, we recommend the following steps:
...
+-  Configure PBS to manage GPU resources.
+
+   Determined works best when allocating GPUs. By default, Determined selects compute nodes with
+   GPUs using the option ``-select={slots_per_trial}:ngpus=1``. If PBS cannot be configured to
+   identify GPUs in this manner, specify the :ref:`pbs section <cluster-configuration-slurm>`
+   ``gres_supported`` option to ``false`` when configuring Determined, and it will then be the
+   user's responsibility to ensure that GPUs will be available on nodes selected for the job using
+   other configurations such as targeting a specific resource pool with only GPU nodes, or
+   specifying a PBS constraint in the experiment configuration.
+
```
Comment (tara-det-ai) [inline docs/setup-cluster/slurm/slurm-requirements.rst:240]: Configure PBS to manage GPU resources.

To optimize GPU allocation, Determined automatically selects compute nodes with GPUs by default using the ``-select={slots_per_trial}:ngpus=1`` option. If PBS cannot identify GPUs in this way, set the :ref:`pbs section <cluster-configuration-slurm>`
   ``gres_supported`` option to ``false`` when configuring Determined. In this case, users must ensure GPU availability on nodes by other means, such as targeting GPU-only resource pools, or specifying a PBS constraint in the experiment configuration.

File: docs/setup-cluster/slurm/slurm-requirements.rst [hunk 2]
```diff
@@ -228,6 +228,94 @@ interacts with PBS, we recommend the following steps:
...
+   PBS should be configured to provide the environment variable ``CUDA_VISIBLE_DEVICES``
+   (``ROCR_VISIBLE_DEVICES`` for ROCm) using a PBS cgroup hook as described in the PBS
+   Administrator's Guide. If PBS is not configured to set ``CUDA_VISIBLE_DEVICES``, Determined will
+   utilize a single GPU on each node. To fully utilize multiple GPUs, you must either manually
+   define ``CUDA_VISIBLE_DEVICES`` appropriately or provide the ``pbs.slots_per_node`` setting in
+   your experiment configuration to indicate how many GPU slots are intended for Determined to use.
+
```
Comment (tara-det-ai) [inline docs/setup-cluster/slurm/slurm-requirements.rst:247]: PBS should be configured to set the environment variable ``CUDA_VISIBLE_DEVICES``
   (or ``ROCR_VISIBLE_DEVICES`` for ROCm) using a PBS cgroup hook, as explained in the PBS
   Administrator's Guide. If PBS is not configured to set ``CUDA_VISIBLE_DEVICES``, Determined will
   utilize only a single GPU on each node. To fully utilize multiple GPUs, you must either manually
   configure ``CUDA_VISIBLE_DEVICES`` or set the ``pbs.slots_per_node`` setting in
   your experiment configuration file to indicate the desired number of GPU slots for Determined.

File: docs/setup-cluster/slurm/slurm-requirements.rst [hunk 3]
```diff
@@ -228,6 +228,94 @@ interacts with PBS, we recommend the following steps:
...
+-  Ensure ``ngpus`` resource is defined with appropriate values.
+
```
Comment (tara-det-ai) [inline docs/setup-cluster/slurm/slurm-requirements.rst:249]: - Ensure the ``ngpus`` resource is defined with the correct values.

File: docs/setup-cluster/slurm/slurm-requirements.rst [hunk 4]
```diff
@@ -228,6 +228,94 @@ interacts with PBS, we recommend the following steps:
...
+   It is necessary to define the ``ngpus`` resource value for every node on the cluster for
+   DeterminedAI to work successfully. Additionally, the resource should also have appropriate flags
+   for PBS to process the value properly while scheduling jobs. Use the following command to check
+   if ``ngpus`` resource is defined with appropriate flags.
+
```
Comment (tara-det-ai) [inline docs/setup-cluster/slurm/slurm-requirements.rst:254]: To ensure the successful operation of Determined, define the ``ngpus`` resource value for each node on the cluster. Additionally, the resource should have the appropriate flags to enable proper processing by PBS when scheduling jobs. You can check if the ``ngpus`` resource is defined with the appropriate flags using the following command:

File: docs/setup-cluster/slurm/slurm-requirements.rst [hunk 5]
```diff
@@ -228,6 +228,94 @@ interacts with PBS, we recommend the following steps:
...
+      [~]$ qmgr -c "list resource ngpus"
+      Resource ngpus
+          type = long
+          flag = hn
+
+   The above output indicates that ngpus is defined as a type ``long`` resource with flags ``h``
+   (indicating this is a host level resource) and ``n`` (indicating this is a consumable resource).
+   If the nodes on your hpc cluster do not provide the same output, use the following commands to
+   create and set the resource ``ngpus`` for every node.
+
```
Comment (tara-det-ai) [inline docs/setup-cluster/slurm/slurm-requirements.rst:266]: The output should indicate that ``ngpus`` is defined as a type ``long`` resource with flags ``h``
   (host-level resource) and ``n`` (consumable resource).
   If your HPC cluster nodes do not produce the same output, use the following commands to
   create and set the resource ``ngpus`` for each node:

File: docs/setup-cluster/slurm/slurm-requirements.rst [hunk 6]
```diff
@@ -228,6 +228,94 @@ interacts with PBS, we recommend the following steps:
...
+   .. code:: bash
+
+      [~]$ qmgr -c "create resource ngpus type=long, flag=hn" # create the ngpus resource
+      [~]$ qmgr -c "set node <nodename> ngpus=<number of GPUs>" # set the value for ngpus
+
+   If you are using virtual nodes (vnodes), ensure ``ngpus`` value is only set on the vnodes and not
+   the actual parent node. Below are the commands and sample output to ensure the same.
+
```
Comment (tara-det-ai) [inline docs/setup-cluster/slurm/slurm-requirements.rst:274]: If you use virtual nodes (vnodes), make sure the ``ngpus`` value is set only on the vnodes, not
   the parent node. Below are the commands and sample output to ensure this:

File: docs/setup-cluster/slurm/slurm-requirements.rst [hunk 7]
```diff
@@ -228,6 +228,94 @@ interacts with PBS, we recommend the following steps:
...
+          resources_available.host = node002
+          resources_available.hpmem = 0b
+          resources_available.mem = 0b
+          resources_available.ncpus = 0 # ngpus value is not set on parent node node002
+          resources_available.Qlist = gpuQ,gpu_hi_priQ
+          resources_available.vmem = 0b
+          resources_available.vnode = node002
+
+   If the ``ngpus`` value is set on the parent node, use the below command to unset it.
+
```
Comment (tara-det-ai) [inline docs/setup-cluster/slurm/slurm-requirements.rst:300]: If the ``ngpus`` value is set on the parent node, use the following command to unset it:

File: docs/setup-cluster/slurm/slurm-requirements.rst [hunk 8]
```diff
@@ -228,6 +228,94 @@ interacts with PBS, we recommend the following steps:
...
+   .. code:: bash
+
+      [madagund@node003 ~]$ sudo qmgr -c "unset node <node_name> resources_available.ngpus"
+
+   The next step is to ensure that ``ngpus`` is listed as a resource in the
+   ``<sched_priv_directory>/sched_config`` file.
+
```
Comment (tara-det-ai) [inline docs/setup-cluster/slurm/slurm-requirements.rst:307]: Next, make sure that ``ngpus`` is listed as a resource in the ``<sched_priv_directory>/sched_config`` file.

File: docs/setup-cluster/slurm/slurm-requirements.rst [hunk 9]
```diff
@@ -228,6 +228,94 @@ interacts with PBS, we recommend the following steps:
    define ``CUDA_VISIBLE_DEVICES`` appropriately or provide the ``pbs.slots_per_node`` setting in
    your experiment configuration to indicate how many GPU slots are intended for Determined to use.
 
+-  Configure PBS to manage GPU resources.
+
```
Comment (tara-det-ai) [inline docs/setup-cluster/slurm/slurm-requirements.rst:232]: When I preview the PR, it looks like there is a duplicate of this section.

**Review by tara-det-ai — COMMENTED** [review]
requested edits, check for duplicates

File: docs/setup-cluster/slurm/slurm-requirements.rst (hunk 9 above)
Comment (jagadeesh545) [reply docs/setup-cluster/slurm/slurm-requirements.rst]: deleted duplicate section

**Comment by jagadeesh545** [comment]
Updated text as suggested and removed the duplicate section.

**Pull request merged**